APP-955 - Flag `untrusted-env` to disable all processes and shell

AdamMagaluk opened the pull request:
For less trusted environments where an untrusted user may update the config and we cannot trust them to run arbitrary code.

Note: This shouldn't be relied upon for completely untrusted environments.

edaniels commented on config/config.go:
```diff
@@ -37,6 +37,11 @@ type Config struct {
+	// DisableProcesses is used to disable Processes for untrusted environments where
+	// a process cannot be trusted. This is an option a user must pass
+	// via command line arguments.
+	DisableProcesses bool `json:"-"`
```
can we make this a broader untrusted flag? this should also disable the shell service as well and we can match on an untrusted env

AdamMagaluk replied:
Yeah that makes sense. Though we should be careful about advertising too much trust at this point.

AdamMagaluk replied:
What do you mean "and we can match on an untrusted env"

edaniels replied:
oh as in just consulting the config and looking for untrusted to disable various features. bad wording :)

AdamMagaluk replied:
Ah yeah gotcha. So far that's shell and processes.

edaniels commented on robot/impl/resource_manager.go:
```diff
@@ -716,26 +724,30 @@ func (manager *resourceManager) updateResources(
-		if oldProc, ok := manager.processManager.RemoveProcessByID(p.ID); ok {
-			if err := oldProc.Stop(); err != nil {
-				manager.logger.Errorw("couldn't stop process", "process", p.ID, "error", err)
+	if !manager.opts.disableProcesses {
```
can we just error out fast if someone does this? it seems like more a violation than skipping things

edaniels commented:
I feel like error out and exit unless that's too harsh

On Monday, November 14, 2022, Adam Magaluk ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In robot/impl/resource_manager.go
> <https://github.com/viamrobotics/rdk/pull/1593#discussion_r1022124791>:
>
> > @@ -716,26 +724,30 @@ func (manager *resourceManager) updateResources(
>  	if err := cleanAppImageEnv(); err != nil {
>  		manager.logger.Errorw("error cleaning up app image environement", "error", err)
>  	}
> -	for _, p := range config.Added.Processes {
> -		_, err := manager.processManager.AddProcessFromConfig(ctx, p)
> -		if err != nil {
> -			manager.logger.Errorw("error while adding process, skipping", "process", p.ID, "error", err)
> -			continue
> -		}
> -	}
> -	for _, p := range config.Modified.Processes {
> -		if oldProc, ok := manager.processManager.RemoveProcessByID(p.ID); ok {
> -			if err := oldProc.Stop(); err != nil {
> -				manager.logger.Errorw("couldn't stop process", "process", p.ID, "error", err)
> +	if !manager.opts.disableProcesses {
>
> What do you expect to happen? RDK exists with non-exit code? i noticed
> errors returned from the resource manager just log to error and don't stop
> RDK.
>
> So if we do this and exit early we log error but also don't setup other
> things like the default services.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/viamrobotics/rdk/pull/1593#discussion_r1022124791>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAGU2QUKW5XRWZOWNAY5ONTWIK32DANCNFSM6AAAAAASAIM6WM>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

AdamMagaluk commented:
> I feel like error out and exit unless that's too harsh
> […](#)
> On Monday, November 14, 2022, Adam Magaluk ***@***.***> wrote: ***@***.**** commented on this pull request. ------------------------------ In robot/impl/resource_manager.go <[#1593 (comment)](https://github.com/viamrobotics/rdk/pull/1593#discussion_r1022124791)>: > @@ -716,26 +724,30 @@ func (manager *resourceManager) updateResources( if err := cleanAppImageEnv(); err != nil { manager.logger.Errorw("error cleaning up app image environement", "error", err) } - for _, p := range config.Added.Processes { - _, err := manager.processManager.AddProcessFromConfig(ctx, p) - if err != nil { - manager.logger.Errorw("error while adding process, skipping", "process", p.ID, "error", err) - continue - } - } - for _, p := range config.Modified.Processes { - if oldProc, ok := manager.processManager.RemoveProcessByID(p.ID); ok { - if err := oldProc.Stop(); err != nil { - manager.logger.Errorw("couldn't stop process", "process", p.ID, "error", err) + if !manager.opts.disableProcesses { What do you expect to happen? RDK exists with non-exit code? i noticed errors returned from the resource manager just log to error and don't stop RDK. So if we do this and exit early we log error but also don't setup other things like the default services. — Reply to this email directly, view it on GitHub <[#1593 (comment)](https://github.com/viamrobotics/rdk/pull/1593#discussion_r1022124791)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAGU2QUKW5XRWZOWNAY5ONTWIK32DANCNFSM6AAAAAASAIM6WM> . You are receiving this because you commented.Message ID: ***@***.***>

My point was that the resource manager reconfiguration errors don't currently force an exit. They just log error level. Changing this behavior now or having a 1 off for this type of error may be a bigger change then we want right now.

edaniels commented:
oh got it. yeah in that case just ignore with error is fine

edaniels reviewed (approved):
LGTM! Thanks for that change
